site/learn/expert-advice/security/breached-password-detection.md

```diff
@@ -0,0 +1,155 @@
+
+> These rules provide less benefit than might be expected because users tend to use predictable methods for satisfying these requirements when imposed (e.g., appending a ! to a memorized secret when required to use a special character). The frustration they often face may also cause them to focus on minimally satisfying the requirements rather than devising a memorable but complex secret. 
+
+Hopefully you're convinced. Now, let's return to our previous discussion of implementation details.
+
+## Finding compromised passwords
+
+First, you need to find the passwords. Have I Been Pwned is a good place to start, but there are other providers out there, such as DeHashed or GhostProject. There may be substantial overlap between these providers. 
+
+You can also include lists of common passwords in your datasets. These may not be present in any public data breach, but are used often enough that they are easy for an attacker to guess, and so should be avoided. Feel free to augment your datasets with any other dictionary lists you can put together. 
+
+Whatever you do, please ensure you include 'correcthorsebatterystaple' ([image courtesy of xkcd](https://xkcd.com/license.html)).
+
+{% include _image.liquid src="/assets/img/advice/breached-password-detection/password-strength.png" alt="Password strength" class="img-fluid text-center" figure=false %}
+
+Make sure you are researching these providers and datasets on a regular schedule, since new breaches happen and new providers may appear. You'll also want to comply with any licensing or other requirements the data providers have. 
+
+Some providers offer these datasets for free, while others may charge or ask for a donation. Make sure you set aside some budget for this.
+
+When you have found these datasets, you'll want to download them, process them, and store it. Again, for a starting point, you can rely on a site like Have I Been Pwned. Make sure you subscribe to their updates.
+
+## Download and store the breached password datasets
+
+Now that you've catalogued the common passwords and known compromised credentials lists, you can build a system to download and store these data sets. Then you'll want to make them available to your authentication system. 
+
+This is a fairly standard data ingestion problem which has been solved in many contexts, so I won't go into nitty gritty details on how to accomplish this. However, make sure you're ingesting this data on a regular schedule. You'll also want to expose this service to your application or applications, possibly using a REST API or data export.
+
+Some services, such as the aforementioned Have I Been Pwned, also offer an HTTP API. Use these APIs if it fits your application needs. In general storing the datasets will give you more control and flexibility. Like any software system, in building breached password detection, you will need to make tradeoffs. 
```

Comment:
I believe the rest of this section is only about "Third-party APIs" but it isn't completely clear. We should make it very clear that the trade-offs are for using APIs you can't control.

Also, it might be good to separate this section into 2 sub-sections:

1. Build your own API
2. Use a third-party API

The first section needs to talk about the time it takes to implement and ensuring that the API calls are very-secure, since you have to send every password to the API. We can also present how FusionAuth does it here as our "light" pitch. We encrypt the password on the client side using a per-client shared secret. We also send it over TLS for two layers of encryption. Then the server decrypts it and processes it. Super-secure.

The second section can discuss the tradeoffs including quicker time to market, reduced cost to implement a homegrown solution, but less control.